### PR TITLE
Feature: Add Relation Filter Type to Custom Reports

### PR DESCRIPTION
**Pull request by gjorgic**
## Changes in this pull request  
Introduce new feature to Custom Reports

## Additional info
user will be able some column mark as relation (_document, asset or object_) and will have ability to filter on same way he can on object grid depending on relation type;
When column is marked as relation, data for that column will be resolved by id and element path will be shown

image

_Future improvement can have some CustomReportShownInGrid interface for object attribute which will display formatted name_

### CODE REVIEW

**Comment by fashxp** [comment]
@gjorgic thx very much. could you please have a look at phpstan?

**Comment by gjorgic** [comment]
I found one issue with opening filter when column is not visible; i will fix today

**Comment by gjorgic** [comment]
now is ready :tada:

File: bundles/CustomReportsBundle/src/Tool/Adapter/Sql.php [hunk 1]
```diff
@@ -46,11 +58,90 @@ public function getData(?array $filters, ?string $sort, ?string $dir, ?int $offs
...
+
+            $this->loadRelationData($data);
         }
 
         return ['data' => $data, 'total' => $total];
     }
 
+    protected function loadRelationData(array &$data): void
+    {
+        $columnsDictionary = $this->getRelationColumns();
```
Comment (fashxp) [inline bundles/CustomReportsBundle/src/Tool/Adapter/Sql.php]: if this is empty, we could skip the whole function, right?

**Review by fashxp — CHANGES_REQUESTED** [review]

File: bundles/CustomReportsBundle/src/Tool/Adapter/Sql.php (hunk 1 above)
Comment (fashxp) [reply bundles/CustomReportsBundle/src/Tool/Adapter/Sql.php]: @gjorgic :ping_pong:

**Comment by gjorgic** [comment]
I have read the CLA Document and I hereby sign the CLA

File: bundles/CustomReportsBundle/src/Tool/Adapter/Sql.php (hunk 1 above)
Comment (gjorgic) [reply bundles/CustomReportsBundle/src/Tool/Adapter/Sql.php]: please, can you check now

File: bundles/CustomReportsBundle/public/js/pimcore/report/custom/report.js [hunk 2]
```diff
@@ -78,8 +95,15 @@ pimcore.bundle.customreports.custom.report = Class.create(pimcore.bundle.customr
             }
 
             if(colConfig["filter"]) {
-                gridColConfig["filter"] = colConfig["filter"];
-                this.gridfilters[colConfig["name"]] = colConfig["filter"];
+                if (colConfig["filter"][0] != '@') {
+                    gridColConfig["filter"] = colConfig["filter"];
+                    this.gridfilters[colConfig["name"]] = colConfig["filter"];
+                } else {
+                    var relationFilterKey = colConfig["filter"].match(this.RELATION_FILTER_REGEX);
```
Comment (fashxp) [inline bundles/CustomReportsBundle/public/js/pimcore/report/custom/report.js:102]: please use `const` or `let` where possible for new things...

File: bundles/CustomReportsBundle/src/Tool/HumanReadableElementNameInterface.php [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+interface HumanReadableElementNameInterface
+{
+    /**
+     * Return a human readable element name
+     *
+     * Allow developers to define what will be returned in the report
+     *
+     * @return string
+     */
+    public function getHumanReadableElementName(): string;
```
Comment (fashxp) [inline bundles/CustomReportsBundle/src/Tool/HumanReadableElementNameInterface.php:15]: we should add some documentation about that feature and how to use it, see also https://pimcore.com/docs/platform/Pimcore/Tools_and_Features/Custom_Reports/

**Review by fashxp — CHANGES_REQUESTED** [review]
thx very much for the contribution. 
sadly, there are some things which do not work yet, especially with the many-to-many filters. 

when setting up a many-to-many object relation filter, I get following error message while opening the report (same for many-to-many asset relation): 

```
URL: /admin/bundle/customreports/custom-report/data?xaction=read&_dc=1713528851388
Method: POST
Message: Pimcore\Bundle\CustomReportsBundle\Tool\Adapter\Sql::loadElementById(): Argument #1 ($id) must be of type int, string given, called in /var/www/html/dev/pimcore/pimcore/bundles/CustomReportsBundle/src/Tool/Adapter/Sql.php on line 92
Trace: 
in /var/www/html/dev/pimcore/pimcore/bundles/CustomReportsBundle/src/Tool/Adapter/Sql.php:127
```
also I'm not sure, if the many-to-many filters work at all. at least in my tests, they didn't filter correctly. 

having a look at the implementation, I'm not too sure if that can work at all?!? 

when opening the many-to-many object relation modal, the field name says `undefined`: 
image


maybe it would make sense to limit that filter feature for many-to-one fields for now?

**Comment by fashxp** [comment]
@gjorgic did you already had a chance to have a look at my comments? Thx very much!

**Comment by gjorgic** [comment]
Hi @fashxp yes, i go through comment, but i Will not be able make update untill end of next week.

Regarding many to many relation, is there chance that issue is related to custom sql query and casting columns to int? If that cause issue do you have any preference how it should be handeled? 

Regarding JS i want keep writing style as is, but ofc, i Can change var to let

**Comment by fashxp** [comment]
@gjorgic thx for you reply, no rush! 

Regarding many-to-many: the thing is, that many-to-many is stored as csv in the query table, e.g. `,34,54,23,` ... so you can't apply an `IN` condition to it, right? Or do I miss something?

**Comment by fashxp** [comment]
@gjorgic :ping_pong:

**Comment by fashxp** [comment]
@gjorgic :ping_pong:

**Comment by fashxp** [comment]
Closing due to inactivity. If you should have time to continue to the discussion just let us know.

**Pull request closed without merging**